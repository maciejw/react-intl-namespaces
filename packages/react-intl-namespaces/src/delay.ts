export interface Cancelable {
  cancel: () => void;
}

export class CancelablePromise<T> implements Promise<T>, Cancelable {
  public [Symbol.toStringTag]: 'Promise';
  private promise: Promise<T>;
  constructor(
    executor: (
      resolve: (value?: T | PromiseLike<T>) => void,
      reject: (reason?: any) => void,
    ) => void,
    public cancel: () => void,
  ) {
    this.promise = new Promise(executor);
  }

  public then<TResult1 = T, TResult2 = never>(
    onfulfilled?:
      | ((value: T) => TResult1 | PromiseLike<TResult1>)
      | null
      | undefined,
    onrejected?:
      | ((reason: any) => TResult2 | PromiseLike<TResult2>)
      | null
      | undefined,
  ): Promise<TResult1 | TResult2> {
    return this.promise.then(onfulfilled, onrejected);
  }
  public catch<TResult = never>(
    onrejected?:
      | ((reason: any) => TResult | PromiseLike<TResult>)
      | null
      | undefined,
  ): Promise<T | TResult> {
    return this.promise.catch(onrejected);
  }
}

export function delay(timeout: number = 0): Promise<void> & Cancelable {
  let timeoutHandle = 0;

  const p = new CancelablePromise<void>(
    resolve => {
      timeoutHandle = window.setTimeout(resolve, timeout);
    },
    () => window.clearTimeout(timeoutHandle),
  );

  return p;
}
