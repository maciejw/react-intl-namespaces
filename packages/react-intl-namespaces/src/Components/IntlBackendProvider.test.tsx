import { mount } from 'enzyme';
import * as React from 'react';
import {
  defineMessages,
  FormattedMessage,
  IntlBackendProvider,
  IntlNamespaceProvider,
} from 'react-intl-namespaces';
import { delay } from '../delay';

const messages = defineMessages<'property1'>({
  property1: {
    defaultMessage: 'text1 {someValue}',
    description: 'description1',
    id: 'property1',
  },
});

const values = { someValue: 'value1' };

describe('IntlNamespaceProvider', () => {
  const sut = (props: IntlBackendProvider.Props = {}) => (
    <IntlBackendProvider locale="en" {...props}>
      <IntlNamespaceProvider
        namespace="namespace"
        includeNamespace={['anotherNamespace']}
      >
        <div>
          <FormattedMessage
            id="anotherNamespace:property1"
            defaultMessage="test2 {someValue}"
            values={values}
          />
          <FormattedMessage {...messages.property1} values={values} />
        </div>
      </IntlNamespaceProvider>
    </IntlBackendProvider>
  );

  const addMissingMessage = jest.fn();

  const getMessagesFromNamespace = jest.fn(async notification => {
    // await delay();
    notification({
      namespace: 'anotherNamespace',
      resource: {
        property1: 'translated text2 {someValue}',
      },
    });
    notification({
      namespace: 'namespace',
      resource: {
        property1: 'translated text1 {someValue}',
      },
    });
  });

  const ResourceProviderMock = jest.fn<IntlBackendProvider.ResourceProvider>(
    () => ({
      addMissingMessageFactory: jest.fn().mockReturnValue(addMissingMessage),
      getMessagesFromNamespaceFactory: jest
        .fn()
        .mockReturnValue(getMessagesFromNamespace),
    }),
  );

  it('should render with default messages', () => {
    const wrapper = mount(sut());

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render with messages ids', () => {
    const wrapper = mount(sut({ showIds: true }));

    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should render with messages with metadata', () => {
    const wrapper = mount(sut({ includeMetadata: true }));

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should provide messages from backend and show them', async () => {
    const mock = new ResourceProviderMock();
    const wrapper = mount(sut({ locale: 'en', ...mock }));

    await delay();

    expect(wrapper.html()).toMatchSnapshot();
    expect(mock).toMatchSnapshot();
    expect(getMessagesFromNamespace).toMatchSnapshot();
    expect(addMissingMessage).toMatchSnapshot();
  });
});
