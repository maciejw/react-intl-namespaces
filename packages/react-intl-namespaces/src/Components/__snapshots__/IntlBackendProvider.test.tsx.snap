// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntlNamespaceProvider should provide messages from backend and show them 1`] = `"<div><span>test2 value1</span><span>translated text1 value1</span></div>"`;

exports[`IntlNamespaceProvider should provide messages from backend and show them 2`] = `
Object {
  "addMissingMessageFactory": [MockFunction] {
    "calls": Array [
      Array [
        [Function],
      ],
    ],
  },
  "getMessagesFromNamespaceFactory": [MockFunction] {
    "calls": Array [
      Array [
        [Function],
      ],
    ],
  },
}
`;

exports[`IntlNamespaceProvider should provide messages from backend and show them 3`] = `
[MockFunction] {
  "calls": Array [
    Array [
      [Function],
      "namespace",
      Array [
        "anotherNamespace",
      ],
    ],
  ],
}
`;

exports[`IntlNamespaceProvider should provide messages from backend and show them 4`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "defaultMessage": "test2 {someValue}",
        "description": undefined,
        "key": "property1",
        "namespace": "anotherNamespace",
      },
    ],
  ],
}
`;

exports[`IntlNamespaceProvider should render with default messages 1`] = `"<div><span>test2 value1</span><span>text1 value1</span></div>"`;

exports[`IntlNamespaceProvider should render with messages ids 1`] = `"<div><span>[anotherNamespace:property1 (someValue)]</span><span>[namespace:property1 (someValue)]</span></div>"`;

exports[`IntlNamespaceProvider should render with messages with metadata 1`] = `"<div><span data-default-message=\\"test2 {someValue}\\" data-key=\\"property1\\" data-namespace=\\"anotherNamespace\\"><span>test2 value1</span></span><span data-default-message=\\"text1 {someValue}\\" data-description=\\"description1\\" data-key=\\"property1\\" data-namespace=\\"namespace\\"><span>text1 value1</span></span></div>"`;
